### PR TITLE
Add proofs for FreeRTOS_ND

### PR DESCRIPTION
**Pull request by moninom1**
Description
-----------
This test add CBMC proofs for prvProcessICMPMessage_IPv6 also validating these functions : 
prvReturnICMP_IPv6
vNDRefreshCacheEntry
prvCheckWaitingBuffer

Test Steps
-----------
CBMC proof run
```
python prepare.py
make
```

Checklist:
----------

- [x] I have tested my changes. No regression in existing tests.
- [x] I have modified and/or added unit-tests to cover the code changes in this Pull Request.

Related Issue
-----------

By submitting this pull request, I confirm that you can use, modify, copy, and redistribute this contribution, under the terms of your choice.

### CODE REVIEW

File: test/cbmc/proofs/ND/prvProcessICMPMessage_IPv6/ProcessICMPMessage_IPv6_harness.c [hunk 1]
```diff
@@ -0,0 +1,163 @@
...
+}
+
+void harness()
+{
+    NetworkBufferDescriptor_t * pxNetworkBuffer = ensure_FreeRTOS_NetworkBuffer_is_allocated();
+    uint32_t ulLen;
+    NetworkBufferDescriptor_t xLocalBuffer;
+    uint16_t usEthernetBufferSize;
+
+    /* The code does not expect both of these to be equal to NULL at the same time. */
```
Comment (ActoryOu) [inline test/cbmc/proofs/ND/prvProcessICMPMessage_IPv6/ProcessICMPMessage_IPv6_harness.c:125]: What does this both mean? There is only one condition below.

File: test/cbmc/proofs/ND/prvProcessICMPMessage_IPv6/ProcessICMPMessage_IPv6_harness.c [hunk 2]
```diff
@@ -0,0 +1,163 @@
...
+    __CPROVER_assume( pxNetworkBuffer != NULL );
+
+    __CPROVER_assume( ( ulLen >= sizeof( ICMPPacket_IPv6_t ) ) && ( ulLen < ipconfigNETWORK_MTU ) );
+
+    pxNetworkBuffer->xDataLength = ulLen;
+    pxNetworkBuffer->pucEthernetBuffer = safeMalloc( ulLen );
+    __CPROVER_assume( pxNetworkBuffer->pucEthernetBuffer );
```
Comment (ActoryOu) [inline test/cbmc/proofs/ND/prvProcessICMPMessage_IPv6/ProcessICMPMessage_IPv6_harness.c:132]:
```suggestion
    __CPROVER_assume( pxNetworkBuffer->pucEthernetBuffer != NULL );
```
For consistency.

File: test/cbmc/proofs/ND/prvProcessICMPMessage_IPv6/ProcessICMPMessage_IPv6_harness.c [hunk 3]
```diff
@@ -0,0 +1,163 @@
...
+
+    /* The packet must at least be as big as an IPv6 Packet. The size is not
+     * checked in the function as the pointer is stored by the IP-task itself
+     * and therefore it will always be of the required size. */
+    __CPROVER_assume( usEthernetBufferSize >= sizeof( IPPacket_IPv6_t ) );
+
+    /* Add matching data length to the network buffer descriptor. */
+    __CPROVER_assume( xLocalBuffer.xDataLength == usEthernetBufferSize );
+
+    xLocalBuffer.pucEthernetBuffer = malloc( usEthernetBufferSize );
```
Comment (ActoryOu) [inline test/cbmc/proofs/ND/prvProcessICMPMessage_IPv6/ProcessICMPMessage_IPv6_harness.c:153]:
```suggestion
    xLocalBuffer.pucEthernetBuffer = safeMalloc( usEthernetBufferSize );
```

File: test/cbmc/proofs/ND/prvProcessICMPMessage_IPv6/ProcessICMPMessage_IPv6_harness.c [hunk 4]
```diff
@@ -0,0 +1,163 @@
...
+
+    /* Since this pointer is maintained by the IP-task, either the pointer
+     * pxARPWaitingNetworkBuffer will be NULL or xLocalBuffer.pucEthernetBuffer
+     * will be non-NULL. */
+    __CPROVER_assume( xLocalBuffer.pucEthernetBuffer != NULL );
+
+    pxARPWaitingNetworkBuffer = &xLocalBuffer;
```
Comment (ActoryOu) [inline test/cbmc/proofs/ND/prvProcessICMPMessage_IPv6/ProcessICMPMessage_IPv6_harness.c:160]: pxARPWaitingNetworkBuffer is always a valid pointer in this setting. Is it necessary to test the scenario for pxARPWaitingNetworkBuffer is NULL?

Comment (ActoryOu) [reply test/cbmc/proofs/ND/prvProcessICMPMessage_IPv6/ProcessICMPMessage_IPv6_harness.c:160]: We can use `pxARPWaitingNetworkBuffer = safeMalloc(xxx);` to simulate either valid or NULL pointer.

**Review by ActoryOu — CHANGES_REQUESTED** [review]

File: test/cbmc/proofs/ND/prvProcessICMPMessage_IPv6/ProcessICMPMessage_IPv6_harness.c (hunk 2 above)
Comment (moninom1) [reply test/cbmc/proofs/ND/prvProcessICMPMessage_IPv6/ProcessICMPMessage_IPv6_harness.c:132]: Updated

File: test/cbmc/proofs/ND/prvProcessICMPMessage_IPv6/ProcessICMPMessage_IPv6_harness.c (hunk 4 above)
Comment (moninom1) [reply test/cbmc/proofs/ND/prvProcessICMPMessage_IPv6/ProcessICMPMessage_IPv6_harness.c:160]: Updated

File: test/cbmc/proofs/ND/prvProcessICMPMessage_IPv6/ProcessICMPMessage_IPv6_harness.c [hunk 5]
```diff
@@ -0,0 +1,169 @@
...
+    {
+        /* The packet must at least be as big as an IPv6 Packet. The size is not
+         * checked in the function as the pointer is stored by the IP-task itself
+         * and therefore it will always be of the required size. */
+        __CPROVER_assume( usEthernetBufferSize >= sizeof( IPPacket_IPv6_t ) );
+
+        /* Add matching data length to the network buffer descriptor. */
+        __CPROVER_assume( xLocalBuffer.xDataLength == usEthernetBufferSize );
+
+        xLocalBuffer.pucEthernetBuffer = malloc( usEthernetBufferSize );
```
Comment (ActoryOu) [inline test/cbmc/proofs/ND/prvProcessICMPMessage_IPv6/ProcessICMPMessage_IPv6_harness.c:154]: Any reason to keep this using malloc instead of safeMalloc?

**Review by ActoryOu — CHANGES_REQUESTED** [review]

File: test/cbmc/proofs/ND/prvProcessICMPMessage_IPv6/ProcessICMPMessage_IPv6_harness.c (hunk 5 above)
Comment (moninom1) [reply test/cbmc/proofs/ND/prvProcessICMPMessage_IPv6/ProcessICMPMessage_IPv6_harness.c:154]: Let me update.

**Review by shubnil — APPROVED** [review]

**Pull request merged**